// @vonage/cli-plugin-conversations/dist/conversations_base.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const cli_utils_1 = __importDefault(require("@vonage/cli-utils"));
const jwt_1 = require("@vonage/jwt");
const vetch_1 = require("@vonage/vetch");
const fs_1 = require("fs");
const lodash_1 = require("lodash");
const types_1 = require("./types");
class ConversationsCommand extends cli_utils_1.default {
    constructor() {
        super(...arguments);
        this._baseurl = `https://api.nexmo.com/v0.3/conversations`;
        this._userBaseurl = `https://api.nexmo.com/v0.3/users`;
        this._defaultHttpOptions = {
            "method": types_1.HTTPMethods.GET,
            "headers": {},
            'responseType': types_1.ResponseTypes.json
        };
    }
    _generateJWT() {
        return __awaiter(this, void 0, void 0, function* () {
            let app_details_raw = (0, fs_1.readFileSync)(`${process.cwd()}/vonage_app.json`);
            let app_details = (JSON.parse(app_details_raw.toString()));
            this._token = yield (0, jwt_1.tokenGenerate)(app_details.application_id, app_details.private_key);
            return;
        });
    }
    init() {
        const _super = Object.create(null, {
            init: { get: () => super.init }
        });
        return __awaiter(this, void 0, void 0, function* () {
            yield _super.init.call(this);
            yield this._generateJWT();
            return;
        });
    }
    getAllConversations(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}`;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    createConversation(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}`;
            opts['method'] = types_1.HTTPMethods.POST;
            opts['data'] = params;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    getConversationById(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${id}`;
            opts['method'] = types_1.HTTPMethods.GET;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    updateConversation(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${params.conversationID}`;
            opts['method'] = types_1.HTTPMethods.PUT;
            opts['data'] = params;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    deleteConversation(id) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${id}`;
            opts['method'] = types_1.HTTPMethods.DELETE;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    getAllMembersInConversation(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${params.conversationID}/members`;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    getMemberById(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${params.conversationID}/members/${params.memberID}`;
            opts['method'] = types_1.HTTPMethods.GET;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    addMemberToConversation(params) {
        return __awaiter(this, void 0, void 0, function* () {
            let data = {
                "state": "joined",
                "user": {
                    "id": params.userID,
                },
                "channel": {
                    "type": "app",
                },
            };
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${params.conversationID}/members`;
            opts['method'] = types_1.HTTPMethods.POST;
            opts['data'] = data;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    removeMemberFromConversation(params) {
        return __awaiter(this, void 0, void 0, function* () {
            let data = {
                "state": "left"
            };
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._baseurl}/${params.conversationID}/members/${params.memberID}`;
            opts['method'] = types_1.HTTPMethods.PATCH;
            opts['data'] = data;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
    getConversationsByUser(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const opts = (0, lodash_1.merge)({}, this._defaultHttpOptions);
            opts['url'] = `${this._userBaseurl}/${params.userID}/conversations/`;
            opts['method'] = types_1.HTTPMethods.GET;
            opts['headers']['Authorization'] = `Bearer ${this._token}`;
            let response = yield (0, vetch_1.request)(opts);
            return response;
        });
    }
}
exports.default = ConversationsCommand;
ConversationsCommand.flags = Object.assign({}, cli_utils_1.default.flags);
ConversationsCommand.args = [
    ...cli_utils_1.default.args,
];
