// @vonage/cli-plugin-conversations/dist/commands/apps/users/conversations.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const conversations_base_1 = __importDefault(require("../../../conversations_base"));
const cli_ux_1 = __importDefault(require("cli-ux"));
class ConversationUpdate extends conversations_base_1.default {
    run() {
        return __awaiter(this, void 0, void 0, function* () {
            const flags = this.parsedFlags;
            const args = this.parsedArgs;
            let response = yield this.getConversationsByUser(Object.assign(Object.assign({}, args), flags));
            let conversationsList = response.data._embedded.conversations;
            cli_ux_1.default.table(conversationsList, {
                id: {},
                name: {},
                display_name: {},
                image_url: {}
            }, Object.assign({}, flags));
        });
    }
}
exports.default = ConversationUpdate;
ConversationUpdate.description = "";
ConversationUpdate.examples = [];
ConversationUpdate.flags = Object.assign(Object.assign({}, conversations_base_1.default.flags), { 'date_start': command_1.flags.string({ description: '' }), 'date_end': command_1.flags.string({ description: '' }), 'page_size': command_1.flags.string({ description: '' }), 'order': command_1.flags.string({ description: '' }), 'cursor': command_1.flags.string({ description: '' }) });
ConversationUpdate.args = [
    ...conversations_base_1.default.args,
    { name: 'userID', required: false }
];
