// @vonage/cli-plugin-numbers/dist/commands/numbers/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const number_base_1 = __importDefault(require("../../number_base"));
const cli_ux_1 = __importDefault(require("cli-ux"));
class NumberList extends number_base_1.default {
    run() {
        return __awaiter(this, void 0, void 0, function* () {
            const flags = this.parsedFlags;
            let numberData = yield this.getAllNumbers({});
            try {
                cli_ux_1.default.table(numberData.numbers, {
                    country: {},
                    msisdn: {
                        header: "Number"
                    },
                    type: {},
                    features: {
                        get: (row) => row.features.join(',')
                    },
                    app_id: {
                        header: "Application",
                        get: (row) => row.app_id || ""
                    }
                }, Object.assign({}, flags));
            }
            catch (error) {
                this.error('No results found.');
            }
            this.exit();
        });
    }
    catch(error) {
        const _super = Object.create(null, {
            catch: { get: () => super.catch }
        });
        return __awaiter(this, void 0, void 0, function* () {
            return _super.catch.call(this, error);
        });
    }
}
exports.default = NumberList;
NumberList.description = 'manage your Vonage numbers';
NumberList.examples = [
    `vonage number`,
];
NumberList.flags = Object.assign(Object.assign({}, number_base_1.default.flags), cli_ux_1.default.table.flags({
    except: ['columns', 'no-truncate', 'csv']
}));
NumberList.args = [];
