// @vonage/cli-plugin-applications/dist/helpers/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.sanitizeFileName = exports.webhookQuestions = void 0;
const prompts_1 = require("prompts");
// address needs to be verified as a valid http address before proceeding
function webhookQuestions({ name, url = 'https://example.com/webhook_name', method = 'POST' }) {
    return __awaiter(this, void 0, void 0, function* () {
        return yield prompts_1.prompt([{
                type: 'text',
                name: 'address',
                message: `${name} - URL`,
                initial: url
            },
            {
                type: 'select',
                name: 'http_method',
                message: `${name} - Method`,
                choices: [
                    { title: 'GET', value: 'GET' },
                    { title: 'POST', value: 'POST' }
                ],
                initial: method === 'POST' ? 1 : 0
            }]);
    });
}
exports.webhookQuestions = webhookQuestions;
function sanitizeFileName(filename, replacement = "_") {
    let illegalRe = /[\/\?<>\\:\*\|":]/g;
    let controlRe = /[\x00-\x1f\x80-\x9f]/g;
    let reservedRe = /^\.+$/;
    let alphanumbericRe = /[^a-z0-9]/ig;
    let windowsReservedRe = /^(con|prn|aux|nul|com[0-9]|lpt[0-9])(\..*)?$/i;
    let sanitized = filename
        .replace(illegalRe, replacement)
        .replace(controlRe, replacement)
        .replace(reservedRe, replacement)
        .replace(windowsReservedRe, replacement)
        .replace(alphanumbericRe, replacement);
    return sanitized;
}
exports.sanitizeFileName = sanitizeFileName;
