// @vonage/vetch/dist/common.js
"use strict";
// Copyright 2020 Vonage
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResponseTypes = exports.HTTPMethods = exports.VetchPromise = exports.VetchError = void 0;
class VetchError extends Error {
    constructor(message, options, response) {
        super(message);
        this.response = response;
        this.config = options;
        this.code = response.status.toString();
    }
}
exports.VetchError = VetchError;
exports.VetchPromise = Promise;
var HTTPMethods;
(function (HTTPMethods) {
    HTTPMethods["GET"] = "GET";
    HTTPMethods["HEAD"] = "HEAD";
    HTTPMethods["POST"] = "POST";
    HTTPMethods["DELETE"] = "DELETE";
    HTTPMethods["PUT"] = "PUT";
    HTTPMethods["CONNECT"] = "CONNECT";
    HTTPMethods["OPTIONS"] = "OPTIONS";
    HTTPMethods["TRACE"] = "TRACE";
    HTTPMethods["PATCH"] = "PATCH";
})(HTTPMethods = exports.HTTPMethods || (exports.HTTPMethods = {}));
var ResponseTypes;
(function (ResponseTypes) {
    ResponseTypes["json"] = "json";
})(ResponseTypes = exports.ResponseTypes || (exports.ResponseTypes = {}));
